hypothesis-python/src/hypothesis/searchstrategy/strings.py

```diff
@@ -63,19 +60,12 @@ def __init__(self,
                 'combination of arguments: ' + ', '.join(
                     '%s=%r' % arg for arg in arguments if arg[1] is not None)
             )
-        self.intervals = IntervalSet(intervals)
-        if whitelist_characters:
-            self.whitelist_characters = set(whitelist_characters)
-        else:
-            self.whitelist_characters = set()
-        self.zero_point = self.intervals.index_above(ord('0'))
+        self.category_intervals = [IntervalSet(i) for i in intervals]
 
     def do_draw(self, data):
-        i = integer_range(
-            data, 0, len(self.intervals) - 1,
-            center=self.zero_point,
-        )
-        return hunichr(self.intervals[i])
+        category = choice(data, self.category_intervals)
+        i = integer_range(data, 0, len(category) - 1)
+        return hunichr(category[i])
```

Comment:
I think this breaks shrinking from one category to another. For example, tests of the form `assert c >= bound` may now often not shrink to `c = bound` - in some cases this will be because the shrinker couldn't find it, in some cases it may be because it's legitimately no longer a shrink!

There's a generic trick to make this sort of thing work, which is that you wrap your initial selection up in an example that "shrinks open".

So the logic looks like:

1. Pick `None` or a category, so that `None` is the zero value.
2. If you pick a category, pick a value from that category then call data.write with the value converted to something that is of the correct size to be read when the value is None.
3. If you picked None pick a value from the whole range.

With an additional start/stop example around the None or category + draw from category, which helps ensure that it can be zeroed properly.

Shrinking this now becomes a two phase thing:

1. First the shrinker causes the category selection to shrink to None.
2. Now the value in the category can be shrunk to the smallest value.